File: leetcode/ConstructBinaryTreefromInorderandPostorderTraversal/buildTree.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public TreeNode buildTree(int[] inorder, int[] postorder) {
        // Start typing your Java solution below
        // DO NOT write main() function
        int len = inorder.length;
        if (len == 0) return null;
        int x = postorder[len - 1], mid = indexOf(inorder, x);
        TreeNode root = new TreeNode(x);
        root.left = buildTree(Arrays.copyOfRange(inorder, 0, mid), Arrays.copyOfRange(postorder, 0, mid));
        root.right = buildTree(Arrays.copyOfRange(inorder, mid + 1, len), Arrays.copyOfRange(postorder, mid, len - 1));
        return root;
    }
    private int indexOf(int arr[], int x) {
        for (int i = 0; i < arr.length; i++) {
            if (arr[i] == x) return i;
        }
        return -1;
    }
}

File: leetcode/ConstructBinaryTreefromInorderandPostorderTraversal/buildTree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* helper(vector<int>& inorder, int ins, int ine, vector<int>& postorder, int posts, int poste) {
        if (ine - ins != poste - posts || posts == poste) return NULL;
        int rootVal = postorder[poste - 1];
        TreeNode *root = new TreeNode(rootVal);
        int pos = find(inorder.begin() + ins, inorder.begin() + ine, rootVal) - inorder.begin();
        root->left = helper(inorder, ins, pos, postorder, posts, pos - ins + posts);
        root->right = helper(inorder, pos + 1, ine, postorder, pos - ins + posts, poste - 1);
        return root;
    }
    TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder) {
        return helper(inorder, 0, inorder.size(), postorder, 0, postorder.size());
    }
};
